Support Gru for IE

miaobin opened the pull request:
@fujunwei  @mingmingtasd  PTAL, thanks!
Test results:
out/Release/webnn_end2end_tests --gtest_filter="Gru*"
Note: Google Test filter = Gru*
[======] Running 2 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 2 tests from GruTests
[ RUN   ] GruTests.GruWith3BatchSize
[      OK ] GruTests.GruWith3BatchSize (27 ms)
[ RUN   ] GruTests.GruWithoutInitialHiddenState
[      OK ] GruTests.GruWithoutInitialHiddenState (5 ms)
[----------] 2 tests from GruTests (33 ms total)

[----------] Global test environment tear-down
[==========] 2 tests from 1 test suite ran. (33 ms total)
[  PASSED  ] 2 tests.

miaobin commented:
MLGruOptions 'resetAfter' should be set to false. The input 'B' in OpenVINO corresponds to the 'bias' and 'recurrentBias' in WebNN spec. According to the description of OpenVINO, if linear_before_reset (resetAfter) is set to 1, then the shape is [num_directions, 4 * hidden_size] - the sum of biases for z and r gates (weights and recurrence weights), the biases for h gate are placed separately. But there is no clear explanation about he biases for h gate.

mingmingtasd commented on src/webnn_native/openvino/GraphIE.cpp:
```diff
@@ -521,6 +521,101 @@ namespace webnn_native { namespace ie {
+            AddConstantWithGraph<size_t>(precision_e::U64, stepsShape, stepsData);
+        auto hiddenSize = gru->GetHiddenSize();
+        ngraph_node_t* biasNode = nullptr;
+        if (options->bias != nullptr) {
```
How about starting to use n and n++ here just for options? For example: int n = 3;

miaobin replied:
Optimized.

mingmingtasd commented on src/webnn_native/openvino/GraphIE.cpp:
```diff
@@ -521,6 +521,101 @@ namespace webnn_native { namespace ie {
+        ngraph_node_t* gruTransposeNode = nullptr;
+        status = return_sequence ? ngraph_transpose(gruNode, order4DNode, &gruTransposeNode)
+                                 : ngraph_transpose(gruNode, orderNode, &gruTransposeNode);
+        if (status != IEStatusCode::OK) {
```
DAWN_TRY(CheckStatusCode(status, "ngraph transpose"));

miaobin replied:
Done.

mingmingtasd commented on src/webnn_native/openvino/GraphIE.cpp:
```diff
@@ -521,6 +521,101 @@ namespace webnn_native { namespace ie {
+        ngraph_node_t* initialHiddenStateTransposeNode = nullptr;
+        status =
+            ngraph_transpose(initialHiddenStateNode, orderNode, &initialHiddenStateTransposeNode);
+        if (status != IEStatusCode::OK) {
```
DAWN_TRY(CheckStatusCode(status, "ngraph transpose"));

miaobin replied:
Done.

mingmingtasd commented on src/webnn_native/openvino/GraphIE.cpp:
```diff
@@ -521,6 +521,101 @@ namespace webnn_native { namespace ie {
+        n++;
+        ngraph_node_t* inputTransposeNode = nullptr;
+        IEStatusCode status = ngraph_transpose(inputNode, orderNode, &inputTransposeNode);
+        if (status != IEStatusCode::OK) {
```
DAWN_TRY(CheckStatusCode(status, "ngraph transpose"));

miaobin replied:
Done.

mingmingtasd commented on src/webnn_native/openvino/GraphIE.cpp:
```diff
@@ -521,6 +521,101 @@ namespace webnn_native { namespace ie {
+            AddConstantWithGraph<int64_t>(precision_e::I64, {order4D.size()}, order4D);
+
+        ngraph_node_t* gruTransposeNode = nullptr;
+        status = return_sequence ? ngraph_transpose(gruNode, order4DNode, &gruTransposeNode)
```
How about naming order as order3D... And add some annotations here to describe the transpose purpose, for example, " Transpose it from shape {?} to shape{?} because ...." .

miaobin replied:
Explained in detail.

mingmingtasd commented on src/webnn_native/ops/Gru.cpp:
```diff
@@ -0,0 +1,78 @@
+                mOptions.initialHiddenState = nullptr;
+            }
+        }
+        mSteps = steps;
```
move L47,48 to L30: `:mSteps(steps),mHiddenSize (hiddenSize)`

miaobin replied:
Optimized.

mingmingtasd commented on src/webnn_native/ops/Gru.cpp:
```diff
@@ -0,0 +1,78 @@
+        mHiddenSize = hiddenSize;
+        mOptions.resetAfter = options == nullptr ? true : options->resetAfter;
+        mOptions.returnSequence = options == nullptr ? false : options->returnSequence;
+        mOptions.direction =
```
I think L33 has set `mOptions.direction` and `mOptions.layout`, no need to add L51~54

miaobin replied:
Optimized.

mingmingtasd commented on src/webnn_native/ops/Gru.cpp:
```diff
@@ -0,0 +1,78 @@
+             OperandBase* input,
+             OperandBase* weight,
+             OperandBase* recurrentWeight,
+             size_t steps,
```
Should we use size_t here and in webnn.json?

miaobin replied:
Yes, we should use int32_t instead of size_t. Done.

mingmingtasd commented on src/webnn_native/ops/Gru.cpp:
```diff
@@ -0,0 +1,78 @@
+        if (mInputs[2]->Rank() != 3) {
+            return DAWN_VALIDATION_ERROR("Argument recurrentWeight is not a 3D tensor.");
+        }
+
```
You should verify all others attributes, for example:
steps must be greater than 0.
hiddenSize: a long scalar. The value of the third dimension of the cell output tensor shape. 
bias is 2-D...
...

mingmingtasd replied:
The options also should be validated, for example, 
hiddenSize: a long scalar. The value of the third dimension of the cell output tensor shape.
bias is 2-D...

miaobin replied:
Done.

mingmingtasd commented on third_party/openvino/ngraph_c_api/src/ngraph_c_api.cpp:
```diff
@@ -116,6 +116,25 @@ inline ngraph::op::PadType GetAutoPad(ngraph_auto_pad autoPad) {
+      ngraph_direction = ngraph::op::RecurrentSequenceDirection::BIDIRECTIONAL;
+      break;
+    default:
+      assert(0);
```
assert would be removed in release build. Please use DAWN_ASSERTPlease use DAWN_ASSERT(0) instead of assert(0), and could you help to add `break` after L112 for fixing https://github.com/webmachinelearning/webnn-native/issues/67? Thanks!

miaobin replied:
'break' added.

mingmingtasd reviewed (commented):
I add some comments, PTAL, thanks! @miaobin

fujunwei commented on src/webnn_native/ops/Gru.cpp:
```diff
@@ -0,0 +1,78 @@
+            if (options->initialHiddenState != nullptr) {
+                mInputs.push_back(options->initialHiddenState);
+            } else {
+                mOptions.bias = nullptr;
```
L.42 - L.44 is unnecessary that is covered by L.32` mOptions = *options;`, I think you worried about the `Operand `in GruOptions isn't nullptr by default?

miaobin replied:
Optimized.

mingmingtasd replied:
L.43 - L.49 is unnecessary, right?

fujunwei replied:
C++ json generator has generator the nullptr for the optional object, so L.44 - L.48 is unnecessary.

miaobin replied:
Yes, these few lines of code are repetitive and unnecessary. Optimized.

miaobin commented on third_party/openvino/ngraph_c_api/src/ngraph_c_api.cpp:
```diff
@@ -510,4 +529,36 @@ IEStatusCode ngraph_group_convolution(const ngraph_node_t* input,
+                                 const bool return_sequence,
+                                 ngraph_node_t** node) {
+  std::vector<std::string> ngraph_activations = {"sigmoid", "tanh"};
+  const std::vector<float> activations_alpha_vector;
```
Done.

miaobin commented on src/webnn_native/openvino/GraphIE.cpp:
```diff
@@ -521,6 +521,101 @@ namespace webnn_native { namespace ie {
+        int n = 0;
+        auto inputs = gru->Inputs();
+        auto options = gru->GetOptions();
+        // TNC <=> NTC
```
Explained in detail.

miaobin commented on third_party/openvino/ngraph_c_api/src/ngraph_c_api.cpp:
```diff
@@ -510,4 +529,36 @@ IEStatusCode ngraph_group_convolution(const ngraph_node_t* input,
+      linear_before_reset);
+  const auto zero =
+      ngraph::op::Constant::create(ngraph::element::Type_t::f32, {}, {0});
+  auto gru_output =
```
Optimized.

miaobin commented on webnn.json:
```diff
@@ -384,6 +411,18 @@
       },
+      {
+        "name": "gru",
+        "returns": "operand",
```
Done.

miaobin commented on webnn.json:
```diff
@@ -384,6 +411,18 @@
+          {"name": "input", "type": "operand"},
+          {"name": "weight", "type": "operand"},
+          {"name": "recurrentWeight", "type": "operand"},
+          {"name": "steps", "type": "size_t"},
```
Done.

miaobin commented on webnn.json:
```diff
@@ -384,6 +411,18 @@
+          {"name": "weight", "type": "operand"},
+          {"name": "recurrentWeight", "type": "operand"},
+          {"name": "steps", "type": "size_t"},
+          {"name": "hiddenSize", "type": "size_t"},
```
Done.

miaobin commented on src/webnn_native/ops/Gru.h:
```diff
@@ -0,0 +1,59 @@
+
+      private:
+        GruOptions mOptions;
+        size_t mSteps;
```
Done.

miaobin commented on src/webnn_native/ops/Gru.h:
```diff
@@ -0,0 +1,59 @@
+      private:
+        GruOptions mOptions;
+        size_t mSteps;
+        size_t mHiddenSize;
```
Done.

fujunwei commented:
@miaobin  please rebase the commit, thanks.

miaobin commented on webnn.json:
```diff
@@ -219,6 +234,18 @@
+      {"name": "reset after", "type": "bool", "default": "true"},
+      {"name": "return sequence", "type": "bool", "default": "false"},
+      {"name": "direction", "type": "recurrent network direction", "default": "forward"},
+      {"name": "layout", "type": "recurrent network weight layout", "default": "zrn"}
```
Support fused activations now.

miaobin commented:
@fujunwei @mingmingtasd Rebased the commit.  PTAL, thanks!

mingmingtasd commented on src/webnn_native/GraphBuilder.cpp:
```diff
@@ -117,6 +118,16 @@ namespace webnn_native {
 
+    OperandArrayBase* GraphBuilderBase::Gru(OperandBase* input,
+                                            OperandBase* weight,
+                                            OperandBase* recurrentWeight,
```
L124 vs L125, int32_t?

miaobin replied:
Optimized.

mingmingtasd commented on src/webnn_native/GraphBuilder.h:
```diff
@@ -43,6 +43,12 @@ namespace webnn_native {
+        OperandArrayBase* Gru(OperandBase*,
+                              OperandBase*,
+                              OperandBase*,
+                              size_t steps,
```
L49 vs L50, int32_t?

miaobin replied:
Optimized.

mingmingtasd commented on third_party/openvino/ngraph_c_api/src/ngraph_c_api.cpp:
```diff
@@ -119,6 +119,48 @@ inline ngraph::op::PadType GetAutoPad(ngraph_auto_pad autoPad) {
   return auto_pad;
 }
 
+inline std::string GetGruActivation(ngraph_gru_activation activation) {
```
Will it bring build warning of uninitialized varibles?

miaobin replied:
Removed this function.

mingmingtasd commented on third_party/openvino/ngraph_c_api/src/ngraph_c_api.cpp:
```diff
@@ -119,6 +119,48 @@ inline ngraph::op::PadType GetAutoPad(ngraph_auto_pad autoPad) {
+  return activations;
+}
+
+inline ngraph::op::RecurrentSequenceDirection get_recurrent_sequence_direction(
```
Will it bring build warning of uninitialized varibles?

miaobin replied:
Removed this function.

fujunwei commented on src/webnn_native/openvino/GraphIE.cpp:
```diff
@@ -574,6 +592,114 @@ namespace webnn_native { namespace ie {
+                                resetActivation, newActivation, linear_before_reset, &gruNode0);
+        DAWN_TRY(CheckStatusCode(status, "ngraph gru"));
+
+        uint32_t ngraph_output_number;
```
Remove it if it's not used.

miaobin replied:
Done.

miaobin commented on third_party/openvino/ngraph_c_api/src/ngraph_c_api.cpp:
```diff
@@ -510,4 +529,36 @@ IEStatusCode ngraph_group_convolution(const ngraph_node_t* input,
+                                 const bool linear_before_reset,
+                                 const bool return_sequence,
+                                 ngraph_node_t** node) {
+  std::vector<std::string> ngraph_activations = {"sigmoid", "tanh"};
```
Supported.

fujunwei commented on webnn.json:
```diff
@@ -242,6 +264,19 @@
+  "gru options": {
+    "category": "structure",
+    "members": [
+      {"name": "bias", "type": "operand"},
```
Add `"optional": true` for bias, recurrent bias, initial hidden state.

miaobin replied:
Optimized.

mingmingtasd commented on third_party/openvino/ngraph_c_api/src/ngraph_c_api.cpp:
```diff
@@ -119,6 +119,48 @@ inline ngraph::op::PadType GetAutoPad(ngraph_auto_pad autoPad) {
+  }
+}
+
+inline std::vector<std::string> GetGruActivations(
```
I think this function is unnecessary, how about removing?

miaobin replied:
Done.

fujunwei commented on third_party/openvino/ngraph_c_api/src/ngraph_c_api.cpp:
```diff
@@ -704,4 +726,50 @@ IEStatusCode ngraph_group_convolution_backprop_data(
+                                 const ngraph_node_t* bias,
+                                 const size_t hidden_size,
+                                 ngraph_recurrent_sequence_direction direction,
+                                 ngraph_gru_activation reset_activation,
```
ngraph is use std::string, so maybe we need to make it as `char*`.

miaobin replied:
This is a good suggestion, I will use char* arrays and delete enum "ngraph_gru_activation".

mingmingtasd commented on third_party/openvino/ngraph_c_api/src/ngraph_c_api.cpp:
```diff
@@ -704,4 +726,50 @@ IEStatusCode ngraph_group_convolution_backprop_data(
+                                 const bool linear_before_reset,
+                                 ngraph_node_t** node) {
+  // activations = {"sigmoid", "tanh"};
+  std::vector<std::string> activations = GetGruActivations(reset_activation, new_activation);
```
`std::vector<std::string> activations = {reset_activation, new_activation};`

miaobin replied:
Now using char* arrays instead of enum "ngraph_gru_activation".

fujunwei reviewed (commented):
Well done, please take a look below comments.

miaobin commented on examples/SampleUtils.cpp:
```diff
@@ -125,6 +128,9 @@ namespace utils {
                 activationOperand = builder.Sigmoid(input);
                 break;
+            case FusedActivation::TANH:
+                activationOperand = builder.Sigmoid(input);
```
Complete the modification.

miaobin commented on examples/SampleUtils.cpp:
```diff
@@ -125,6 +128,9 @@ namespace utils {
                 activationOperand = builder.Sigmoid(input);
                 break;
+            case FusedActivation::TANH:
+                activationOperand = builder.Sigmoid(input);
```
Done.

miaobin commented on src/webnn_native/openvino/GraphIE.cpp:
```diff
@@ -134,6 +134,24 @@ namespace webnn_native { namespace ie {
             return status;
         }
 
+        ngraph_gru_activation getGruActivation(OperatorBase* gruOperator) {
```
Done.

fujunwei commented:
@miaobin please rebase it and fix the code format issue, thanks.

miaobin commented on src/tests/end2end/GruTests.cpp:
```diff
@@ -0,0 +1,96 @@
+    }
+
+  public:
+    void TestGru(const std::vector<int32_t>& inputShape,
```
Optimized.

miaobin commented on third_party/openvino/ngraph_c_api/src/ngraph_c_api.cpp:
```diff
@@ -510,4 +529,36 @@ IEStatusCode ngraph_group_convolution(const ngraph_node_t* input,
+                                 ngraph_node_t** node) {
+  std::vector<std::string> ngraph_activations = {"sigmoid", "tanh"};
+  const std::vector<float> activations_alpha_vector;
+  const std::vector<float> activations_beta_vector;
```
Discussed.

miaobin commented:
- Optimized and rebased the code.
- Added check for all parameters.
- Added test cases.
- Updated the test results.
@fujunwei @mingmingtasd PTAL, thanks a lot!

mingmingtasd commented on src/webnn_native/openvino/GraphIE.cpp:
```diff
@@ -622,6 +640,119 @@ namespace webnn_native { namespace ie {
         return {};
     }
 
+    MaybeError Graph::AddGru(const op::Gru* gru) {
```
Need you consider the case for options->layout = rzn?

Need you check for `direction: a MLRecurrentNetworkDirection. The processing direction of the input sequence. When set to "both", the size of the first dimension of the weight and the bias tensor shapes must be 2, and the input is processed in both directions.` ?

miaobin replied:
Added the check code.

mingmingtasd commented on src/webnn_native/ops/Gru.cpp:
```diff
@@ -0,0 +1,97 @@
+                return DAWN_VALIDATION_ERROR("Argument recurrentBias is not a 2D tensor.");
+            }
+        }
+        // The initialHiddenState 2-D tensor
```
// The initialHiddenState 3-D tensor

miaobin replied:
Done.

miaobin commented on src/tests/end2end/GruTests.cpp:
```diff
@@ -0,0 +1,155 @@
+        utils::BuildConstant(builder, initialHiddenStateShape, initialHiddenStateData.data(),
+                             initialHiddenStateData.size() * sizeof(float));
+
+    const ml::RecurrentNetworkDirection direction = ml::RecurrentNetworkDirection::Forward;
```
Removed the unnecessary code.

miaobin commented on src/webnn_native/openvino/GraphIE.cpp:
```diff
@@ -622,6 +640,119 @@ namespace webnn_native { namespace ie {
+            AddConstantWithGraph<int64_t>(precision_e::I64, {order3D.size()}, order3D);
+        // [steps, num_directions, batch_size, hidden_size] => [batch_size, num_directions, steps,
+        // hidden_size]
+        std::vector<int64_t> order4D = std::vector<int64_t>{2, 0, 1, 3};
```
Optimized.

miaobin commented on src/webnn_native/openvino/GraphIE.cpp:
```diff
@@ -622,6 +640,119 @@ namespace webnn_native { namespace ie {
+            ngraph_transpose(initialHiddenStateNode, order3DNode, &initialHiddenStateTransposeNode);
+        DAWN_TRY(CheckStatusCode(status, "Transpose gru initialHiddenState layout"));
+        auto direction = static_cast<ngraph_recurrent_sequence_direction>(options->direction);
+        // TODO: layout
```
Added the assert.

fujunwei reviewed (approved):
Well done, LGTM.

mingmingtasd commented:
Great work, LGTM, thanks! @miaobin

fujunwei commented:
Merge it.
